### PR TITLE
Add unit tests for `DefaultQubit._apply_operation` and move gate integration tests to device test suite

### PR DESCRIPTION
**Pull request by antalszava**
Changes testing of `DefaultQubit` the following way:

* Runs the device test suite for `default.qubit`
* Moves the gate-specific tests from the `TestDefaultQubitIntegration` class of `DefaultQubit` to the device test suite
* Removes some wires tests that are identical in the device test suite
* Adds unit tests for the `_apply_operation` `DefaultQubit` method that test that the control flow (the order in which statements are executed) is as expected

Note: with more tests, the runtime for the `test_gates.py` file increased slightly, although not significantly. For `default.qubit` a local increase of `0.97s  ->  1.48s`, for `qiskit.aer` an increase of `2.56s  -> 5.08s` was observed. This means, that on the CI, the increased runtime is still negligible compared to dependency installation (which usually takes ~25 sec).

### CODE REVIEW

**Comment by antalszava** [comment]
[ch4085]

File: pennylane/devices/tests/test_gates.py
Comment (antalszava) [inline pennylane/devices/tests/test_gates.py:375]: codefactor caught that this was defined in the outer scope

File: pennylane/devices/tests/test_gates.py
Comment (antalszava) [inline pennylane/devices/tests/test_gates.py:609]: All of these tests were adopted from the tests for `DefaultQubit` (which are later removed in this PR).

The added value in having these tests is that we now have cases which test gate applications with `qml.expval`.

File: tests/devices/test_default_qubit.py
Comment (antalszava) [inline tests/devices/test_default_qubit.py:711]: found these in the device test suite

File: tests/devices/test_default_qubit.py
Comment (antalszava) [inline tests/devices/test_default_qubit.py:774]: moved these to the device test suite

File: tests/devices/test_default_qubit.py
Comment (antalszava) [inline tests/devices/test_default_qubit.py:897]: took the liberty of abandoning this test case (quite many other basis state tests)

File: tests/devices/test_default_qubit.py
Comment (antalszava) [inline tests/devices/test_default_qubit.py:1733]: the remaining tests are present in the test suite

File: tests/devices/test_default_qubit.py
Comment (antalszava) [inline tests/devices/test_default_qubit.py:1665]: These tests are new, they test the `_apply_operation` method

File: .github/workflows/tests.yml
Comment (josh146) [inline .github/workflows/tests.yml:87]: Nice!

Don't forget to also update `codecov.yml`, to increase the `after_n_builds` parameter.

File: pennylane/devices/tests/test_gates.py
Comment (josh146) [reply pennylane/devices/tests/test_gates.py:609]: Nice!

While potentially redundant at the moment, since `QubitDevice` uses probabilities + eigenvalues to compute expectation values, this is definitely worth having, since we could have devices that use separate logic for expectation values.

Comment (josh146) [reply pennylane/devices/tests/test_gates.py:609]: Do you think it's worth having this in a separate file? This file is becoming quite large.

File: tests/devices/test_default_qubit.py
Comment (josh146) [reply tests/devices/test_default_qubit.py:1665]: amazing!

File: tests/devices/test_default_qubit.py
Comment (josh146) [inline tests/devices/test_default_qubit.py:1688]: Docstring could potentially be made more specific

File: tests/devices/test_default_qubit.py
Comment (josh146) [inline tests/devices/test_default_qubit.py:1700]: Should this be a case where the `monkeypatching` fixture is used?

File: tests/devices/test_default_qubit.py
Comment (josh146) [inline tests/devices/test_default_qubit.py:1742]: same question here re: monkeypatching

**Review by josh146 — APPROVED** [review]
This is a great addition @antalszava, and help removes a lot of technical debt!

I approve, but left some thoughts and suggestions.

File: .github/workflows/tests.yml
Comment (antalszava) [reply .github/workflows/tests.yml:87]: Hmm, what does `after_n_builds` do? :thinking: (incremented it)

File: pennylane/devices/tests/test_gates.py
Comment (antalszava) [reply pennylane/devices/tests/test_gates.py]: Definitely, that's a great point :) moved them to a new test in d8cb74fd931c13831cbae8ab812fe35b7e894641

File: tests/devices/test_default_qubit.py
Comment (antalszava) [reply tests/devices/test_default_qubit.py]: Good point, thank you! Updated these.

File: tests/devices/test_default_qubit.py
Comment (antalszava) [reply tests/devices/test_default_qubit.py]: Updated

File: .github/workflows/tests.yml
Comment (josh146) [reply .github/workflows/tests.yml:87]: It essentially tells Codecov not to send the coverage report to GitHub until `n` CI builds have been performed. This stops Codecov from sending incorrect coverage updates when not all tests have yet been run.

I see that you've incremented it to 8, but you'll need to increment it to 9, since you are adding two new CI tests here :)

**Review by josh146 — APPROVED** [review]
Thanks @antalszava! Approving, but let's not merge until after the release.

File: pennylane/devices/tests/test_gates_with_expval.py
Comment (mariaschuld) [inline pennylane/devices/tests/test_gates_with_expval.py:73]: clever test cases!

File: pennylane/devices/tests/test_gates_with_expval.py
Comment (mariaschuld) [inline pennylane/devices/tests/test_gates_with_expval.py:140]:
```suggestion
        """Tests supported parametrized gates that act on a single wire"""
```

File: pennylane/devices/tests/test_gates_with_expval.py
Comment (mariaschuld) [inline pennylane/devices/tests/test_gates_with_expval.py:153]:
```suggestion
    # This test uses initial state 1/2|00>+sqrt(3)/2|11> and checks two Z expvals
```

File: pennylane/devices/tests/test_gates_with_expval.py
Comment (mariaschuld) [inline pennylane/devices/tests/test_gates_with_expval.py:208]: Check the docsting, I am not sure if I understand correctly, are the gates parametrised?

File: pennylane/devices/tests/test_gates_with_expval.py
Comment (mariaschuld) [reply pennylane/devices/tests/test_gates_with_expval.py:153]: Update here and everywhere else?

File: pennylane/devices/tests/test_gates_with_expval.py
Comment (mariaschuld) [reply pennylane/devices/tests/test_gates_with_expval.py:140]: Update here and everywhere else?

File: tests/devices/test_default_qubit.py
Comment (mariaschuld) [inline tests/devices/test_default_qubit.py:1671]: Nice! I would find the test clearer if it would use a name like `SupportedGate` instead of `PauliX`...it took me a while to understand that you redefine that gate!

File: tests/devices/test_default_qubit.py
Comment (mariaschuld) [inline tests/devices/test_default_qubit.py:1721]:
```suggestion
        # Recreate the S gate, so that it is an example for a one-qubit gate
        # that does not inherit from DiagonalOperation
```

File: tests/devices/test_default_qubit.py
Comment (mariaschuld) [inline tests/devices/test_default_qubit.py:1714]: docstring does not match test name?

File: tests/devices/test_default_qubit.py
Comment (mariaschuld) [inline tests/devices/test_default_qubit.py:1756]: Same here...

File: tests/devices/test_default_qubit.py
Comment (mariaschuld) [inline tests/devices/test_default_qubit.py:1762]:
```suggestion
        # Redefine the Toffoli gate so that it is an example for a gate with
```

Comment (mariaschuld) [reply tests/devices/test_default_qubit.py:1762]: I am a bit confused why a gate with two wires would use tensordot?

**Review by mariaschuld — CHANGES_REQUESTED** [review]
Only cosmetic changes. Thanks for this great cleanup!

File: .github/workflows/tests.yml
Comment (antalszava) [reply .github/workflows/tests.yml:87]: Oh nice! Sure, updated now :)

File: pennylane/devices/tests/test_gates_with_expval.py
Comment (antalszava) [reply pennylane/devices/tests/test_gates_with_expval.py:73]: courtesy of the `default.qubit` test suite :slightly_smiling_face:

File: tests/devices/test_default_qubit.py
Comment (antalszava) [reply tests/devices/test_default_qubit.py:1671]: Oh, good point! changed to have `supported_gate_application` as the new method and also added monkeypatching to make it a cleaner solution. (Note that the gate itself is not redefined, rather the internal function that `default.qubit` uses to apply `PauliX` is, adding this to the docstring.)

File: tests/devices/test_default_qubit.py
Comment (antalszava) [reply tests/devices/test_default_qubit.py:1714]: Oh good catch, thanks! :slightly_smiling_face:

File: tests/devices/test_default_qubit.py
Comment (antalszava) [reply tests/devices/test_default_qubit.py:1762]: It's due to speed that was observed previously, `_apply_unitary` looks like this:
```python
        if isinstance(operation, DiagonalOperation):
            return self._apply_diagonal_unitary(state, matrix, wires)
        if len(wires) <= 2:
            # Einsum is faster for small gates
            return self._apply_unitary_einsum(state, matrix, wires)

        return self._apply_unitary(state, matrix, wires)
```
This test tests the last `return` statement.

**Comment by antalszava** [comment]
Thanks for the comments @mariaschuld, addressed them :slightly_smiling_face:

**Review by mariaschuld — APPROVED** [review]

**Pull request merged**
